core/TypePtr.h

```diff
@@ -13,6 +13,23 @@ class TypeConstraint;
 struct DispatchResult;
 struct DispatchArgs;
 
+class NonRefcounted {};
+
+class Refcounted {
+    friend class TypePtrTestHelper;
+    std::atomic<uint32_t> counter{0};
+
+public:
+    void addref() {
+        this->counter.fetch_add(1);
+    }
+
+    uint32_t release() {
+        // fetch_sub returns value prior to subtract
+        return this->counter.fetch_sub(1) - 1;
+    }
+};
```

Comment:
idk if you care about having a `CheckSize` here. It seems maybe useful as a point of documentation that this must be 4 bytes so that the other half of the 8 bytes can fit a NameRef